File: core/x/javaapi/api/src/main/java/viaduct/java/api/variables/VariablesProvider.java
package viaduct.java.api.variables;

import java.util.Map;
import java.util.concurrent.CompletableFuture;
import viaduct.java.api.context.VariablesProviderContext;
import viaduct.java.api.types.Arguments;

/**
 * Java equivalent of Kotlin's {@code VariablesProvider}. Implementations dynamically compute values
 * for GraphQL variables that are referenced from a resolver's required selection set fragment.
 *
 * <p>An implementation must be a nested static class on a resolver and must be annotated with
 * {@link viaduct.java.api.annotations.Variables} declaring the variable names and types produced by
 * {@link #provide}.
 *
 * <p>The returned map must contain exactly the keys declared in the {@code @Variables} annotation.
 *
 * @param <A> The arguments type for the field whose resolver references this provider's variables.
 *     Use {@link Arguments.NoArguments} when the field has no arguments.
 */
@FunctionalInterface
public interface VariablesProvider<A extends Arguments> {

  /**
   * Compute and return the variable values.
   *
   * @param context Provides access to the field arguments and request context.
   * @return A future containing the resolved variable values, keyed by variable name.
   */
  CompletableFuture<Map<String, Object>> provide(VariablesProviderContext<A> context);
}

File: core/tenant/api/src/main/kotlin/viaduct/api/resolver/VariablesProvider.kt
package viaduct.api.resolver

import viaduct.api.context.VariablesProviderContext
import viaduct.api.types.Arguments
import viaduct.apiannotations.StableApi

/**
 * A base interface for dynamic provisioning of variable values.
 * All implementations of VariablesProvider must be annotated with [Variables].
 *
 * If a viaduct resolver declares a VariablesProvider, it will be invoked at request time to
 * provide additional variable values that may be used in an [Resolver.objectValueFragment].
 *
 * Example:
 * ```kotlin
 * @Resolver("baz(${'$'}x)")
 * class FooBarResolver : FooResolvers.Bar() {
 *
 *   @Variables("x: Int", "y: String")
 *   class Vars : VariablesProvider<Arguments.NoArguments> {
 *     override suspend fun provide(context: VariablesProviderContext<Arguments.NoArguments>): Map<String, Any?> =
 *       mapOf("x" to 42)
 *   }
 * }
 * ```
 */
@StableApi
fun interface VariablesProvider<A : Arguments> {
    /**
     * Return a Map with provided variable values.
     * The map must contain exactly the keys in the [Variables] annotation on the containing
     * VariablesProvider class.
     */
    suspend fun provide(context: VariablesProviderContext<A>): Map<String, Any?>
}
